### PR TITLE
chore(v2): document use of aws-crt-client (#1092)

### PR DESCRIPTION
**Pull request by jreijn**
**Issue #, if available:** #1092

## Description of changes:

- Added documentation for using the aws-crt-client with powertools (v2).
- Fixed some documentation issues with typos and code blocks.

**Checklist**

 

* [ ] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda-java/#tenets)
* [ ] Update tests
* [x] Update docs
* [ ] PR title follows [conventional commit semantics](https://www.conventionalcommits.org/en/v1.0.0/)

## Breaking change checklist

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/FAQs.md [hunk 1]
```diff
@@ -47,3 +47,91 @@ To enable `forceAjcCompile` you need to use following `aspectj-maven-plugin` con
...
+    <version>1.18.0</version>
+    <exclusions>
+        <exclusion>
+            <groupId>software.amazon.awssdk</groupId>
+            <artifactId>url-connection-client</artifactId>
+        </exclusion>
+    </exclusions>
+</dependency>
+```
+Next to adding the `aws-crt-client` and excluding the `aws-crt` dependency (contain all runtime libraries) it's required to set a specific runtime version of the `aws-crt` dependency by specifying the classifier for your specific target runtime.
```
Comment (jeromevdl) [inline docs/FAQs.md:73]:
```suggestion
Next, add the `aws-crt-client` and exclude the "generic" `aws-crt` dependency (contains all runtime libraries). 
Instead, set a specific classifier of the `aws-crt` to use the one for your target runtime: either `linux-x86_64` for a Lambda configured for x86 or `linux-aarch_64` for Lambda using arm64.
```

File: docs/FAQs.md [hunk 2]
```diff
@@ -47,3 +47,91 @@ To enable `forceAjcCompile` you need to use following `aspectj-maven-plugin` con
...
+Specifying the specific target runtime makes sure all other target runtimes are excluded from the jar file, which will result in the benefit of improved cold start times.
```
Comment (jeromevdl) [inline docs/FAQs.md:74]: not sure to understand that sentence, if we exclude the generic crt, what else do we need to exclude?

Comment (jeromevdl) [inline docs/FAQs.md:74]: Understood, then maybe rephrase:
```suggestion
By specifying the specific target runtime, it will avoid other target runtimes to be included in the lambda jar file, which will result in a smaller package and the benefit of improved cold start times.   
```

File: docs/FAQs.md [hunk 3]
```diff
@@ -47,3 +47,91 @@ To enable `forceAjcCompile` you need to use following `aspectj-maven-plugin` con
...
+    <dependency>
+        <groupId>software.amazon.awssdk.crt</groupId>
+        <artifactId>aws-crt</artifactId>
+        <version>0.29.9</version>
+        <classifier>linux-x86_64</classifier>
+    </dependency>
+</dependencies>
+```
+
+After configuring the dependencies it's required to specify the aws sdk http client.
```
Comment (jeromevdl) [inline docs/FAQs.md:100]:
```suggestion
After configuring the dependencies it's required to specify the AWS SDK http client in the code. 
```

File: docs/FAQs.md [hunk 4]
```diff
@@ -47,3 +47,91 @@ To enable `forceAjcCompile` you need to use following `aspectj-maven-plugin` con
...
+        <groupId>software.amazon.awssdk.crt</groupId>
+        <artifactId>aws-crt</artifactId>
+        <version>0.29.9</version>
+        <classifier>linux-x86_64</classifier>
+    </dependency>
+</dependencies>
+```
+
+After configuring the dependencies it's required to specify the aws sdk http client. 
+Most modules support a custom sdk client by leveraging the `.withClient()` method on the for instance the Provider singleton:
```
Comment (jeromevdl) [inline docs/FAQs.md:101]: something missing here...

Comment (jreijn) [reply docs/FAQs.md]: I've tried to clarify it a bit more. I noticed I focussed too much on the parameters module, but other modules also use the url-connection client. Turned it into an example for a module and described that it depends on the PT module in use how to further configure the http client.

**Comment by jreijn** [comment]
@jeromevdl thank you for the thorough review! I've incorporated most of your suggestions!

File: docs/FAQs.md [hunk 5]
```diff
@@ -47,3 +47,97 @@ To enable `forceAjcCompile` you need to use following `aspectj-maven-plugin` con
 </configuration>
 ```
 
+## How can I use Powertools for AWS Lambda (Java) with the AWS CRT HTTP Client?
+
+Powertools uses the `url-connection-client` as the default http client. The `url-connection-client` is a lightweight http client, which keeps the impact on Lambda cold starts to a minimum.
```
Comment (scottgerring) [inline docs/FAQs.md:52]:
```suggestion
Powertools uses the `url-connection-client` as the default HTTP client. The `url-connection-client` is a lightweight HTTP client, which keeps the impact on Lambda cold starts to a minimum. 
```

File: docs/FAQs.md [hunk 6]
```diff
@@ -47,3 +47,97 @@ To enable `forceAjcCompile` you need to use following `aspectj-maven-plugin` con
 </configuration>
 ```
 
+## How can I use Powertools for AWS Lambda (Java) with the AWS CRT HTTP Client?
+
+Powertools uses the `url-connection-client` as the default http client. The `url-connection-client` is a lightweight http client, which keeps the impact on Lambda cold starts to a minimum. 
+With the [announcement](https://aws.amazon.com/blogs/developer/announcing-availability-of-the-aws-crt-http-client-in-the-aws-sdk-for-java-2-x/) of the `aws-crt-client` a new http client has been released, which offers faster SDK startup time and smaller memory footprint.
```
Comment (scottgerring) [inline docs/FAQs.md:53]:
```suggestion
With the [announcement](https://aws.amazon.com/blogs/developer/announcing-availability-of-the-aws-crt-http-client-in-the-aws-sdk-for-java-2-x/) of the `aws-crt-client` a new HTTP client has been released, which offers faster SDK startup time and smaller memory footprint. 
```

File: docs/FAQs.md [hunk 7]
```diff
@@ -47,3 +47,97 @@ To enable `forceAjcCompile` you need to use following `aspectj-maven-plugin` con
...
+        <exclusion>
+            <groupId>software.amazon.awssdk</groupId>
+            <artifactId>url-connection-client</artifactId>
+        </exclusion>
+    </exclusions>
+</dependency>
+```
+Next, add the `aws-crt-client` and exclude the "generic" `aws-crt` dependency (contains all runtime libraries). 
+Instead, set a specific classifier of the `aws-crt` to use the one for your target runtime: either `linux-x86_64` for a Lambda configured for x86 or `linux-aarch_64` for Lambda using arm64.
+By specifying the specific target runtime, it will avoid other target runtimes to be included in the jar file, which will result in a smaller Lambda package and the benefit of improved cold start times.
```
Comment (scottgerring) [inline docs/FAQs.md:77]:
```suggestion
By specifying the specific target runtime, we prevent other target runtimes from being included in the jar file, resulting in a smaller Lambda package and improved cold start times.
```

File: docs/FAQs.md [hunk 8]
```diff
@@ -47,3 +47,97 @@ To enable `forceAjcCompile` you need to use following `aspectj-maven-plugin` con
...
+        <groupId>software.amazon.awssdk.crt</groupId>
+        <artifactId>aws-crt</artifactId>
+        <version>0.29.9</version>
+        <classifier>linux-x86_64</classifier>
+    </dependency>
+</dependencies>
+```
+
+### Explicitly set the AWS CRT HTTP Client
+After configuring the dependencies, it's required to explicitly specify the AWS SDK http client.
```
Comment (scottgerring) [inline docs/FAQs.md:104]:
```suggestion
After configuring the dependencies, it's required to explicitly specify the AWS SDK HTTP client. 
```

File: docs/FAQs.md [hunk 9]
```diff
@@ -47,3 +47,97 @@ To enable `forceAjcCompile` you need to use following `aspectj-maven-plugin` con
...
+        <artifactId>aws-crt</artifactId>
+        <version>0.29.9</version>
+        <classifier>linux-x86_64</classifier>
+    </dependency>
+</dependencies>
+```
+
+### Explicitly set the AWS CRT HTTP Client
+After configuring the dependencies, it's required to explicitly specify the AWS SDK http client. 
+Depending on the Powertools module, there is a different way to configure the sdk client.
```
Comment (scottgerring) [inline docs/FAQs.md:105]:
```suggestion
Depending on the Powertools module, there is a different way to configure the SDK client.
```

File: docs/FAQs.md [hunk 10]
```diff
@@ -47,3 +47,97 @@ To enable `forceAjcCompile` you need to use following `aspectj-maven-plugin` con
...
+                    .get("/my/secret");
+                    // We might instead want to retrieve multiple parameters at once, returning a Map of key/value pairs
+                    // .getMultiple("/my/secret/path");
+
+            // Return the result
+            return value;
+        }
+    }
+    ```
+It has been considered to make the `aws-crt-client` the default http client in Lambda Powertools for Java, as mentioned in [Move SDK http client to CRT](https://github.com/aws-powertools/powertools-lambda-java/issues/1092),
```
Comment (scottgerring) [inline docs/FAQs.md:142]:
```suggestion
The `aws-crt-client` was considered for adoption as the default HTTP client in Lambda Powertools for Java as mentioned in [Move SDK http client to CRT](https://github.com/aws-powertools/powertools-lambda-java/issues/1092), 
```

**Review by scottgerring — CHANGES_REQUESTED** [review]
Nice! Couple of nitpicks inline but looking good. Also immediately helpful for the next time I need to remember how to set this all up :D

File: docs/FAQs.md [hunk 11]
```diff
@@ -47,3 +47,97 @@ To enable `forceAjcCompile` you need to use following `aspectj-maven-plugin` con
 </configuration>
 ```
 
+## How can I use Powertools for AWS Lambda (Java) with the AWS CRT HTTP Client?
+
+Powertools uses the `url-connection-client` as the default http client. The `url-connection-client` is a lightweight http client, which keeps the impact on Lambda cold starts to a minimum. 
+With the [announcement](https://aws.amazon.com/blogs/developer/announcing-availability-of-the-aws-crt-http-client-in-the-aws-sdk-for-java-2-x/) of the `aws-crt-client` a new http client has been released, which offers faster SDK startup time and smaller memory footprint. 
+
+Unfortunately, replacing the `url-connection-client` dependency with the `aws-crt-client` will not immediately improve the lambda cold start performance and memory footprint, 
+as the default version of the dependency contains low level libraries for all target runtimes (Linux, MacOS, Windows, etc).
```
[the remark this answers is not in the document]
Comment (jreijn) [reply docs/FAQs.md:56]: Yes definitely great improvement.

**Comment by jreijn** [comment]
Some great additions @scottgerring ! Thanks so much. I do have one outstanding issue for myself. I mentioned the  `<version>1.18.0</version>` as the version for the dependency. As this is going into the v2 branch should I replace that with 2.x.x? As the code example is for the v2 branch.

**Comment by scottgerring** [comment]
> Some great additions @scottgerring ! Thanks so much. I do have one outstanding issue for myself. I mentioned the `<version>1.18.0</version>` as the version for the dependency. As this is going into the v2 branch should I replace that with 2.x.x? As the code example is for the v2 branch.

If you use 2.0.0 we should be safe! We've got a rather rough script in the build that goes through and subs this all out when we do releases. I believe we're using 2.0.0 elsewhere in the docs.

**Review by scottgerring — APPROVED** [review]
LGTM! Cheers @jreijn
